File: drills/elixir/solutions/nc44_gas_station.exs
defmodule Solution do
  def can_complete_circuit([], _cost), do: -1

  # Two facts do all the work. If the total gas is short of the total cost no
  # start works at all; and if the tank runs dry between i and j, no station in
  # between can start either, so the search jumps straight to j + 1 rather than
  # restarting.
  def can_complete_circuit(gas, cost) do
    {total, _tank, start} =
      gas
      |> Enum.zip(cost)
      |> Enum.with_index()
      |> Enum.reduce({0, 0, 0}, fn {{g, c}, i}, {total, tank, start} ->
        diff = g - c
        tank = tank + diff
        if tank < 0, do: {total + diff, 0, i + 1}, else: {total + diff, tank, start}
      end)

    if total >= 0, do: start, else: -1
  end
end

File: drills/go/solutions/nc44_gas_station.go
package main

func canCompleteCircuit(gas []int, cost []int) int {
	total, tank, start := 0, 0, 0
	for i := range gas {
		total += gas[i] - cost[i]
		tank += gas[i] - cost[i]
		// Running dry here means no start between the last reset and here
		// can work either: they would all arrive with even less.
		if tank < 0 {
			start = i + 1
			tank = 0
		}
	}
	if total < 0 {
		return -1
	}
	return start
}
